only return distinct paths from map matching top k

kevinkreiser opened the pull request:
fixes #970 

basically it does the simplest thing which is sticks the results in a map and has the map do the logic of saying whether or not a particular sequence of edges is different from another. ive also added the score to the map using accumulated cost. this is not robust to discontinuities so it definitely needs some work but its a good first start until @ptpt has his alternate approach ready.

kevinkreiser commented:
having some more thoughts about this. it might not work when intermediate candidates switch between node snapped and not. see i took a shortcut and tried to just use the matchresult edgeids as a proxy for the path but the problem is if a match result is bouncing between an edge and a node then the edges might be off by one. the only way to really do this for real is to recover the path which was something i wanted to avoid doing... i'll have a look closer to see..

kevinkreiser commented:
ok so ive refactored it to not just look at edge ids as a proxy for the path by making it not optional to actually compute the path from the match set. currently this isnt a big deal as no code paths were such that this didnt happen but its possible for a feature in the future not to need this. until such a time comes that we dont need the path then we can leave it this way.

there is still one issue im working on which is how to terminate the top k loop when we have enough unique results. it seems to me that sometimes you cant get `k` unique results and so we have to detect that and terminate with `< k` results. trying to knock this out right now

ptpt commented:
What about adding a param to `ViterbiSearch::SearchPath(Time time, bool break=False)`? If `break` is true, it tells VS to break if a state disconnects with its previous state.

kevinkreiser commented:
yeah i was pitching that idea awhile ago and kind of forgot about it! that sounds like the easiest approach. and i mean who cares what the difference is between cost of one disconnected path vs another at that point! thanks for reminding me again @ptpt !

ptpt commented on src/meili/map_matcher.cc:
```diff
@@ -417,19 +422,51 @@ MapMatcher::OfflineMatch(const std::vector<Measurement>& measurements, uint32_t
+    match_results.score = accumulated_cost;
+
+    // Keep this path if its unique
+    auto hash = hasher(match_results);
```
Is it possible to move this part code to a function `IsSamePath(vector<MatchResult>& rhs, vector<MatchResult>& rhs)` to reduce the complexity of OfflineMatch?

kevinkreiser replied:
sure thing!

ptpt replied:
Also, I assume that if the current path is different from the previous best path, then it should be different from all previous best paths. With this assumption, we don't need to compare it with all previous best paths. WDYT?

kevinkreiser replied:
are you saying we only need to compare to the previous best path and not keep all the hashes? i had the same thought at first but then i was worried that perhaps you could have a simple case like in the issue this references, where its just a simple fork in the road and the first result takes the northern edge of the fork, the second result takes the lower edge of the fork and then the 3rd result node snaps and takes the top edge of the fork again.

also although the code is somewhat complicated in the end im only storing a hash and an index, so there is very little overhead. its basically no extra memory or computation to have all of the paths' hashes to check against. i think the extra protection is worth it because i dont think i can prove that we dont need to account of it picking the same result that we found earlier than the last result. i'm making more test cases and we'll see if maybe we can prove this could be simplified.

kevinkreiser commented:
ok @ptpt i think ive addressed all your comments. have another look. i just need to get the logic finished to detect that special case i put on the end of the #970

kevinkreiser commented:
ok ive made some more changes to facilitate fixing that last case added to the end of #970. but it brings me back around to the comment that @ptpt made about how many to check. so even for a short trace like the one in the test that @dgearhart made, which has only 6 points, the combinations of possibilities that exist are large, somewhere in the 10's of thousands. even if you set `k=2` for this test, it takes many seconds to complete because it has to go through all of those thousands of possibilities just to find out that they are all the same and really there is only 1 unique answer. this makes me think that maybe if we see after several iterations that we've added no more unique answers we should give up. i am going to test the code to see where the slowness lies but for any kind of reasonably long path the combinatoric nature of possibilities makes this intractable after we start rejecting paths that are basically the same.

dgearhart commented:
:shipit:
